sqlmesh/core/context.py

```diff
@@ -427,12 +426,20 @@ def upsert_model(self, model: t.Union[str, Model], **kwargs: t.Any) -> Model:
             raise SQLMeshError(f"The disabled model '{model.name}' cannot be upserted")
         path = model._path
 
-        # model.copy() can't be used here due to a cached state that can be a part of a model instance.
-        model = t.cast(Model, type(model)(**{**t.cast(Model, model).dict(), **kwargs}))
+        model = model.copy(update=kwargs)
+        model._full_depends_on = None  # the query can be updated to change deps
```

Comment:
Why do we not do this in the model's `copy()`?

Reply:
because most of the times you want to cache _full_depends_on, it's just in context.upsert (like in tests), we mutate the query which can change depends_on

Reply:
I think this may lead to too many gotchas. I'd check in `copy` whether `update` include `depends_on` or `query`. And then invalidate based on that. Otherwise this is too error prone.

Reply:
Also I don't see us resetting renderers on copy

Reply:
happens when you to context.upsert_model, used in tests, and notebooks